### PR TITLE
Fix backward incompatibility in TTL execution

### LINKED ISSUE #63700: Broken moving to S3 by TTL data parts that created in 24.1 to S3 after upgrade to 24.2
> Please make sure that the version you're using is still supported (you can find the list [here](https://github.com/ClickHouse/ClickHouse/blob/master/SECURITY.md#scope-and-supported-versions)).

> You have to provide the following information whenever possible.

I use Clickhouse 24.1
I've created table and filled with data with TTL (move to S3).

Update Clickhouse to 24.2.
Old parts are not moved to S3 after the TTL expires. There are no problem with new parts.

Example:
```
DROP TABLE test_ttl_table;

CREATE TABLE test_ttl_table (
	generation UInt64, 
	date_key DateTime, 
	number UInt64, 
	text String, 
	expired DateTime DEFAULT now()
) 
ENGINE=MergeTree 
ORDER BY (generation, date_key) 
PARTITION BY toMonth(date_key) 
TTL expired + INTERVAL 10 MINUTES TO DISK 'object_storage';


INSERT INTO test_ttl_table (
	generation,
	date_key,
	number,
	text
)
	SELECT
		1,
		toDateTime('2000-01-01 00:00:00') + rand(number) % 365 * 86400,
		number,
		toString(number)

	FROM numbers(10000000);

# update cluster to 24.2

SELECT
    partition,
    name,
    uuid,
    disk_name,
    move_ttl_info.min,
    move_ttl_info.max,
    bytes_on_disk
FROM system.parts
WHERE table = 'test_ttl_table'
ORDER BY toInt8(partition) ASC

Query id: e4bab693-3865-44a3-a2ac-664c0d82103c

┌─partition─┬─name─────────┬─uuid─────────────────────────────────┬─disk_name─┬─move_ttl_info.min───────┬─move_ttl_info.max───────┬─bytes_on_disk─┐
│ 1         │ 1_74_74_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        928770 │
│ 1         │ 1_93_93_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        930934 │
│ 1         │ 1_97_97_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        929958 │
│ 1         │ 1_115_115_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        492933 │
│ 1         │ 1_10_69_1    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5759285 │
│ 2         │ 2_7_62_1     │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5394180 │
│ 2         │ 2_110_110_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        460464 │
│ 2         │ 2_105_105_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        876439 │
│ 2         │ 2_92_92_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        869973 │
│ 2         │ 2_82_82_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        870436 │
│ 3         │ 3_117_117_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        488874 │
│ 3         │ 3_101_101_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        931489 │
│ 3         │ 3_94_94_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        930016 │
│ 3         │ 3_81_81_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        929802 │
│ 3         │ 3_11_67_1    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5770230 │
│ 4         │ 4_6_64_1     │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5580195 │
│ 4         │ 4_111_111_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        475575 │
│ 4         │ 4_106_106_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        901198 │
│ 4         │ 4_91_91_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        900482 │
│ 4         │ 4_79_79_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        900967 │
│ 5         │ 5_116_116_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        490023 │
│ 5         │ 5_104_104_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        926871 │
│ 5         │ 5_87_87_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        928796 │
│ 5         │ 5_84_84_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        932116 │
│ 5         │ 5_8_72_1     │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5771681 │
│ 6         │ 6_107_107_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        900318 │
│ 6         │ 6_120_120_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        470141 │
│ 6         │ 6_89_89_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        900147 │
│ 6         │ 6_80_80_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        899309 │
│ 6         │ 6_9_70_1     │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5579481 │
│ 7         │ 7_12_63_1    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5752767 │
│ 7         │ 7_76_76_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        925182 │
│ 7         │ 7_96_96_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        930183 │
│ 7         │ 7_103_103_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        925495 │
│ 7         │ 7_109_109_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        487546 │
│ 8         │ 8_4_68_1     │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5766840 │
│ 8         │ 8_83_83_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        926365 │
│ 8         │ 8_90_90_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        925285 │
│ 8         │ 8_98_98_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        930100 │
│ 8         │ 8_112_112_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        495023 │
│ 9         │ 9_114_114_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        476192 │
│ 9         │ 9_2_66_1     │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5583211 │
│ 9         │ 9_77_77_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        898147 │
│ 9         │ 9_88_88_0    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        902041 │
│ 9         │ 9_102_102_0  │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        903643 │
│ 10        │ 10_1_65_1    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5764723 │
│ 10        │ 10_75_75_0   │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        933274 │
│ 10        │ 10_86_86_0   │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        927844 │
│ 10        │ 10_99_99_0   │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        928094 │
│ 10        │ 10_118_118_0 │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        487836 │
│ 11        │ 11_5_71_1    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5568108 │
│ 11        │ 11_73_73_0   │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        899391 │
│ 11        │ 11_95_95_0   │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        900161 │
│ 11        │ 11_108_108_0 │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        896461 │
│ 11        │ 11_113_113_0 │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        471726 │
│ 12        │ 12_3_61_1    │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │       5570900 │
│ 12        │ 12_119_119_0 │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        471695 │
│ 12        │ 12_100_100_0 │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        901110 │
│ 12        │ 12_85_85_0   │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        900259 │
│ 12        │ 12_78_78_0   │ 00000000-0000-0000-0000-000000000000 │ default   │ ['2024-05-09 15:08:29'] │ ['2024-05-09 15:08:29'] │        898663 │
└───────────┴──────────────┴──────────────────────────────────────┴───────────┴─────────────────────────┴─────────────────────────┴───────────────┘

60 rows in set. Elapsed: 0.005 sec. 

rc1a-g123qt0h5lfckic3.mdb.yandexcloud.net :) select now()

SELECT now()

Query id: 6208cfb0-6062-4761-95d8-fce51e2e80d7

┌───────────────now()─┐
│ 2024-05-09 15:44:57 │
└─────────────────────┘

1 row in set. Elapsed: 0.002 sec.
```

**Expected behavior**

Expected that parts will move to S3.

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix bug which could lead to non-working TTLs with expressions

Fixes #63700.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit a400911d7a463f4dc09521d34ea669287987a134 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64694/a400911d7a463f4dc09521d34ea669287987a134/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by serxa — APPROVED** [review]

File: tests/integration/test_move_ttl_broken_compatibility/test.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+@pytest.fixture(scope="module")
+def started_cluster():
+    try:
+        cluster.add_instance(
+            "node1",
+            main_configs=["configs/storage_conf.xml"],
+            image="clickhouse/clickhouse-server",
+            with_minio=True,
+            tag="24.1",
```
Comment (maxknv) [inline tests/integration/test_move_ttl_broken_compatibility/test.py:24]: why do we hardcode version? is it one-time test?

File: src/Storages/MergeTree/MergeTreeDataPartTTLInfo.cpp [hunk 2]
```diff
@@ -3,6 +3,9 @@
 #include <IO/WriteHelpers.h>
 #include <Common/quoteString.h>
 #include <algorithm>
+#include <Common/logger_useful.h>
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeDataPartTTLInfo.cpp:6]:
```suggestion
```

File: tests/integration/test_move_ttl_broken_compatibility/test.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        )
+        .strip()
+        .split("\n")
+    )
+    print("Disks before", disks)
+
+    assert len(disks) == 1
+    assert disks[0] == "default"
+
+    node1.restart_with_latest_version()
```
Comment (antonio2368) [inline tests/integration/test_move_ttl_broken_compatibility/test.py:87]: :exploding_head:

**Pull request merged**